packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,382 @@
+import { blockFromRpc } from '@ethereumjs/block/dist/from-rpc'
+import { Trie } from '@ethereumjs/trie'
+import {
+  Account,
+  bigIntToHex,
+  bufferToBigInt,
+  bufferToHex,
+  intToHex,
+  isHexPrefixed,
+  setLengthLeft,
+  toBuffer,
+} from '@ethereumjs/util'
+import { JsonRpcProvider, StaticJsonRpcProvider } from '@ethersproject/providers'
+import { debug } from 'debug'
+import { keccak256 } from 'ethereum-cryptography/keccak'
+
+import { Cache } from './cache'
+
+import { BaseStateManager } from '.'
+
+import type { Proof, StateManager } from '.'
+import type { getCb, putCb } from './cache'
+import type { StorageDump } from './interface'
+import type { Common } from '@ethereumjs/common'
+import type { Address } from '@ethereumjs/util'
+
+const log = debug('statemanager')
+
+type BlockTagType = bigint | 'latest' | 'earliest' | 'pending'
+export interface EthersStateManagerOpts {
+  provider: string | StaticJsonRpcProvider | JsonRpcProvider
+  blockTag?: BlockTagType
+}
+
+export class EthersStateManager extends BaseStateManager implements StateManager {
+  private provider: StaticJsonRpcProvider | JsonRpcProvider
+  private contractCache: Map<string, Buffer>
+  private storageCache: Map<string, Map<string, Buffer>>
+  private blockTag: string
+  private stateRoot: Buffer | undefined
+  _cache: Cache
+
+  constructor(opts: EthersStateManagerOpts) {
+    super({})
+    if (typeof opts.provider === 'string') {
+      this.provider = new StaticJsonRpcProvider(opts.provider)
+    } else if (opts.provider instanceof JsonRpcProvider) {
+      this.provider = opts.provider
+    } else {
+      throw new Error(`valid JsonRpcProvider or url required; got ${opts.provider}`)
+    }
+
+    this.blockTag =
+      typeof opts.blockTag === 'bigint' ? bigIntToHex(opts.blockTag) : opts.blockTag ?? 'latest'
+
+    this.contractCache = new Map()
+    this.storageCache = new Map()
+
+    /*
+     * For a custom StateManager implementation adopt these
+     * callbacks passed to the `Cache` instantiated to perform
+     * the `get`, `put` and `delete` operations with the
+     * desired backend.
+     */
+    const getCb: getCb = async (address) => {
+      return this.getAccountFromProvider(address)
+    }
+    const putCb: putCb = async (_keyBuf, _accountRlp) => {
+      return Promise.resolve()
+    }
+    const deleteCb = async (_keyBuf: Buffer) => {
+      return Promise.resolve()
+    }
+    this._cache = new Cache({ getCb, putCb, deleteCb })
+  }
+
+  copy(): EthersStateManager {
+    const newState = new EthersStateManager({ provider: this.provider })
+    ;(newState as any).contractCache = new Map(this.contractCache)
```

Comment:
Just mentioning some stuff I notice along comparison to old EthersStateManager, not necessarily something to act upon (judge yourself): is the state itself (before: trie) actually copied here?

Reply:
it seems like changes in cache below will actually persist between copes, as well as for storage cache's elements

Reply:
`new Map(x)` where `x` is of type `Map` actually copies the entire map so these are copied. But the `_cache` itself is not copied, so this looks a bit inconsistent?